### PR TITLE
Make constant `threadPool.relativeTimeInMillisSupplier()`

### PR DESCRIPTION
**Pull request by DaveCTurner**
We use `ThreadPool#relativeTimeInMillis` as part of a timeout mechanism
in many places, often using a method reference so we can pass in a
`LongSupplier` for ease of testing. Each time we create a new method
reference we're opening ourselves up to potential cache misses on the
way through to the two volatile reads needed to actually return the
value in production. We're also only caching the nanos value so we must
do some arithmethic to convert it to millis each time.

This commit introduces a constant `LongSupplier` for use in all these
places, saving the allocation and the indirection through different
objects, and caching the millis value alongside the nanos to avoid the
arithmetic on each call.

Relates #104273

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I'd be interested to know if this makes any difference too. In theory it could.

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 1]
```diff
@@ -738,6 +767,20 @@ public void run() {
                 }
             }
         }
+
+        LongSupplier relativeTimeInMillisSupplier() {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:771]: NIT: Kinda weird to have this as a function when we only call it once during the lifetime of any instance, why not just make this a final field that we set up in the constructor of the thread pool? Doesn't really matter much though :)

**Review by original-brownbear — APPROVED** [review]
LGTM, probably not going to change much but still should have zero downside for saving some heap + code cache/cpu-cache saving :)

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:771]: Yeah I can tidy this up; also this area is pretty light on tests right now, I'll add some before marking this as non-draft.

**Pull request merged**